File: leetcode/1026_Maximum_Difference_Between_Node_and_Ancestor.py
from typing import Optional
import unittest

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        diff = 0
        stack = [(root, root.val, root.val)]
        while stack:
            node, min_value, max_value = stack.pop()
            if abs(min_value - node.val) > diff:
                diff = abs(min_value - node.val)
            if abs(max_value - node.val) > diff:
                diff = abs(max_value - node.val)
            if node.val < min_value:
                min_value= node.val
            elif node.val > max_value:
                max_value = node.val
            if node.left is not None:
                stack.append((node.left, min_value, max_value))
            if node.right is not None:
                stack.append((node.right, min_value, max_value))
        return diff

    def _maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        diff = 0
        def _rec(node: Optional[TreeNode], min_value: int, max_value: int):
            nonlocal diff
            if node is not None:
                if abs(min_value - node.val) > diff:
                    diff = abs(min_value - node.val)
                if abs(max_value - node.val) > diff:
                    diff = abs(max_value - node.val)
                if node.val < min_value:
                    min_value= node.val
                elif node.val > max_value:
                    max_value = node.val
                _rec(node.left, min_value, max_value)
                _rec(node.right, min_value, max_value)
        _rec(root, root.val, root.val)
        return diff

class MaxAncestorDiff(unittest.TestCase):
    def test_1(self) -> None:
        self.assertEqual(
            Solution().maxAncestorDiff(
                TreeNode(8, TreeNode(3, TreeNode(1, TreeNode(6, TreeNode(4), TreeNode(7)))), TreeNode(10, None, TreeNode(14, TreeNode(13))))
            ),
            7
        )

    def test_2(self) -> None:
        self.assertEqual(
            Solution().maxAncestorDiff(
                TreeNode(1, None, TreeNode(2, None, TreeNode(0, TreeNode(3))))
            ),
            3
        )

File: leetcode/1026_Maximum_Difference_Between_Node_and_Ancestor.rs
use std::rc::Rc;
use std::cell::RefCell;
use std::cmp::{min, max};

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val: val,
            left: None,
            right: None,
        }
    }
}

struct Solution {}

impl Solution {
    pub fn max_ancestor_diff(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut diff = 0;
        if let Some(node) = root {
            let mut stack = vec![(node.clone(), node.borrow().val, node.borrow().val)];
            while let Some((node, min_value, max_value)) = stack.pop() {
                let val = node.borrow().val;
                let min_value = min(min_value, val);
                let max_value = max(max_value, val);
                diff = max(diff, (min_value - val).abs());
                diff = max(diff, (max_value - val).abs());
                if let Some(left) = &node.borrow().left {
                    stack.push((left.clone(), min_value, max_value));
                }
                if let Some(right) = &node.borrow().right {
                    stack.push((right.clone(), min_value, max_value));              
                }
            }
        }
        diff
    }
}

mod max_ancestor_diff_test {
    use super::*;

    fn node(val: i32, left: Option<Rc<RefCell<TreeNode>>>, right: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode {val, left, right})))
    }

    #[test]
    fn test_1() {
        assert_eq!(
            Solution::max_ancestor_diff(
                node(1, None, node(2, None, node(0, node(3, None, None), None)))
            ), 
            3
        );
    }
}
